### PR TITLE
Fix callback group logic in executor

### LINKED ISSUE #2474: Bug in Executor::get_next_ready_executable ?
Hey

https://github.com/ros2/rclcpp/blob/f9c4894f96ea083fc5acfcea4f1ea83850759e63/rclcpp/src/rclcpp/executor.cpp#L692

I am highly confused by this code block:
```c++
        auto callback_group = entity_iter->second.callback_group.lock();
        if (callback_group && !callback_group->can_be_taken_from()) {
          continue;
        }
``` 

Shouldn't it be ?
```c++
        auto callback_group = entity_iter->second.callback_group.lock();
        if (!callback_group || !callback_group->can_be_taken_from()) {
          continue;
        }
```

### PR DESCRIPTION
**Pull request by mjcarroll**
Fixes #2474

CC: @jmachowinski

### CODE REVIEW

**Review by alsora — APPROVED** [review]
Is it easy to create a unit-test that would fail without this PR and succeeds now?

**Review by jmachowinski — APPROVED** [review]

**Review by fujitatomoya — COMMENTED** [review]
@mjcarroll do we need to apply the same fix for, (where is irrelevant from https://github.com/ros2/rclcpp/pull/2142)

https://github.com/ros2/rclcpp/blob/f9c4894f96ea083fc5acfcea4f1ea83850759e63/rclcpp/src/rclcpp/executors/executor_entities_collection.cpp#L156

`callback_group` is null means that entity is not valid. AFAIS, `rclcpp` does not use  `ready_executables` but user would?

CC: @jmachowinski

**Comment by jmachowinski** [comment]
> Is it easy to create a unit-test that would fail without this PR and succeeds now?
Should be straight forward.
A custom executor, 
Add a cbg
call collect, 
then delete the cbg
Call get next executable.
This failed for me, I got a any executable with an empty cbg

File: rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
+// Copyright 2017 Open Source Robotics Foundation, Inc.
```
Comment (jmachowinski) [inline rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp:1]: C&P ?

**Review by jmachowinski — APPROVED** [review]

File: rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+  rclcpp::AnyExecutable next() {
+    rclcpp::AnyExecutable ret;
+    this->get_next_ready_executable(ret);
+    return ret;
+  }
+};
+
+
+TEST(TestCallbackGroup, ValidCbg)
```
Comment (wjwwood) [inline rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp:62]: nitpick, we typically use snakecase for the test name and I'd appreciate it if we try to use the full name for things (helps with translation for people who don't speak english and pattern matching), e.g.:

```suggestion
TEST(TestCallbackGroup, valid_callback_group)
```

File: rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  EXPECT_EQ(nullptr, next_executable.client);
+  EXPECT_EQ(nullptr, next_executable.service);
+  EXPECT_EQ(nullptr, next_executable.subscription);
+  EXPECT_EQ(nullptr, next_executable.waitable);
+  EXPECT_EQ(nullptr, next_executable.data);
+
+  rclcpp::shutdown();
+}
+
+TEST(TestCallbackGroup, InvalidCbg)
```
Comment (wjwwood) [inline rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp:98]: nitpick:

```suggestion
TEST(TestCallbackGroup, invalid_callback_group)
```

File: rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+{
+  rclcpp::init(0, nullptr);
+
+  // Create a timer associated with a callback group
+  auto node = std::make_shared<rclcpp::Node>("node");
+
+  // Add the callback group to the executor
+  auto executor = CustomExecutor();
+  executor.add_node(node);
+  executor.spin_all(std::chrono::milliseconds(10));
```
Comment (wjwwood) [inline rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp:108]: Just curious why is this needed?

**Review by wjwwood — COMMENTED** [review]
changes lgtm, had a question about the tests

Also a general comment on the test, I'd love to see the use of `spin()` instead of `spin_all()`, or even `spin_once()` in a loop. And I've found that using a custom waitable that can be explicitly triggered rather than a timer can avoid unintentional race conditions and sleeps. Though we've been copying that custom waitable around a lot. It might be a good idea to create a new entity type that is like a `TriggeredCallback` class which would inherit from `Waitable` and would differ from a `GuardCondition` in that it has a callback to be called when the executor executes it, rather than when it is triggered (which the `GuardCondition` class already has). It would definitely be useful in tests and may even be useful for users, allowing them to manually schedule a callback to be run in the executor on demand.

If others agree I could work on that.

File: rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp (hunk 2 above)
Comment (alsora) [reply rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp:62]: According to gtest docs, `PascalCase` should be used

> TEST() arguments go from general to specific. The first argument is the name of the test suite, and the second argument is the test’s name within the test suite. Both names must be valid C++ identifiers, and they should not contain any underscores (_)

from https://google.github.io/googletest/primer.html#simple-tests

**Comment by jmachowinski** [comment]
> And I've found that using a custom waitable that can be explicitly triggered rather than a timer can avoid unintentional race conditions and sleeps

In this special case, the timer is not racy, as we just need it to be ready. Also I would like to keep the timer, as it uncovered a bug causing an endless loop.

File: rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp (hunk 4 above)
Comment (mjcarroll) [reply rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp:108]: This was to clear out all of the node startup transients.

File: rclcpp/src/rclcpp/executor.cpp [hunk 5]
```diff
@@ -689,7 +689,7 @@ Executor::get_next_ready_executable(AnyExecutable & any_executable)
       auto entity_iter = current_collection_.timers.find(timer->get_timer_handle().get());
       if (entity_iter != current_collection_.timers.end()) {
         auto callback_group = entity_iter->second.callback_group.lock();
-        if (callback_group && !callback_group->can_be_taken_from()) {
+        if (!callback_group || !callback_group->can_be_taken_from()) {
```
Comment (jmachowinski) [inline rclcpp/src/rclcpp/executor.cpp:692]: current_timer_index++;

Comment (jmachowinski) [reply rclcpp/src/rclcpp/executor.cpp:692]: Also below

File: rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  node.reset();
+  timer.reset();
+
+  // Since the callback group has been reset, this should not be allowed to
+  // be a valid executable (timer and cbg should both be nullptr).
+  // In the regression, timer == next_executable.timer whil
+  // next_executable.callback_group == nullptr, which was incorrect.
+  executor.wait();
+  auto next_executable = executor.next();
+  EXPECT_NE(nullptr, next_executable.timer);
```
Comment (jmachowinski) [inline rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp:130]: EXPECT_EQ

**Comment by mjcarroll** [comment]
@jmachowinski I have incorporated your changes from: https://github.com/cellumation/rclcpp/tree/fix_cbg_condition

They have been squashed and I marekd you with co-authored-b

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20760/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15059/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21439/)

File: rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp (hunk 2 above)
Comment (wjwwood) [reply rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp:62]: That's interesting, I never knew that, looking into it, it seems like it doesn't matter now, but could in the future: https://stackoverflow.com/a/34543433

I think we should refactor them all to avoid this condition, but obviously not in this pr. @clalancette

Comment (clalancette) [reply rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp:62]: Wow, I had no idea.  We have literally thousands of gtests in ROS 2 using underscore.

That said, I actually think we should stick with our current snake-case style, even though it is slightly wrong.  To me it seems better to be consistent, and if we want to fix this (pretty theoretical) issue, we can do it all at once.

**Comment by jmachowinski** [comment]
@mjcarroll I picked the test over to https://github.com/ros2/rclcpp/pull/2494

2494 also fixes the tests fails that came up in the CI.

I think we can close this PR in favor for 2494

**Comment by wjwwood** [comment]
Ok, let's go ahead and close this then. We can reopen it if needed, and I think the feed back was addressed in the new pr as far as I can tell.

For future reference, it's usually better to make prs to other prs rather than replacing them, though sometimes that doesn't work out due to needing to rebase. Or like me you mess up the pr or branch some how and have to close it 🙃.
